runtime/protocol/http-test/common/src/aws/smithy/kotlin/runtime/httptest/TestConnection.kt

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.httptest
+
+import aws.smithy.kotlin.runtime.http.Headers
+import aws.smithy.kotlin.runtime.http.HttpBody
+import aws.smithy.kotlin.runtime.http.HttpStatusCode
+import aws.smithy.kotlin.runtime.http.engine.HttpClientEngineBase
+import aws.smithy.kotlin.runtime.http.engine.callContext
+import aws.smithy.kotlin.runtime.http.readAll
+import aws.smithy.kotlin.runtime.http.request.HttpRequest
+import aws.smithy.kotlin.runtime.http.request.HttpRequestBuilder
+import aws.smithy.kotlin.runtime.http.response.HttpCall
+import aws.smithy.kotlin.runtime.http.response.HttpResponse
+import aws.smithy.kotlin.runtime.time.Instant
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+/**
+ * An expected HttpRequest with the response that should be returned by the engine
+ * @param request the expected request. If null no assertions are made on the request
+ * @param response the response to return for this request. If null it defaults to an empty 200-OK response
+ */
+data class ExpectedHttpRequest(val request: HttpRequest?, val response: HttpResponse? = null)
+
+/**
+ * Actual and expected [HttpRequest] pair
+ */
+data class CapturedRequest(val expected: HttpRequest?, val actual: HttpRequest) {
+    /**
+     * Assert that all of the components set on [expected] are also the same on [actual]. The actual request
+     * may have additional headers, only the ones set in [expected] are compared.
+     */
+    internal suspend fun assertRequest(idx: Int) {
+        if (expected == null) return
+        assertEquals(expected.url.toString(), actual.url.toString(), "[request#$idx]: URL mismatch")
+        expected.headers.forEach { name, values ->
+            values.forEach {
+                assertTrue(actual.headers.contains(name, it), "[request#$idx]: header $name missing value $it")
```

Comment:
nit

enclose key/values in single quotes to make debugging w/ empty values easier.